### PR TITLE
[Inductor-CPP] If all of the activation scale dims are 1, make it a 0D tensor

### PR DESCRIPTION
**Pull request by sanchitintel**
For int8 dynamically quantized activation & int8 quantized weights, add a workaround for some indexing issue that expected an empty index ( so, was expecting a 0D tensor) in epilogue creator when the activation scale was sized [1, 1] by converting it into a 0D tensor.

The issue was discovered while running LLaMA2 quantized with torchao's `int8_dynamic_activation_int8_weight` quantization on CPU with max-autotune enabled (although this error would've occurred regardless).

The final hidden states tensor that's activation to LM head is of shape `[batch_size, sequence_length, hidden_dim]` during decoding. For decoding one token at a time with batch size 1, sequence length is 1. The activation scale is shaped `[1, 1]` (reshaped from `[1, 1, 1]`). However, Inductor epilogue creator expects a 0D tensor in this case (my guess is that the corresponding logic in Inductor expects a 0D tensor if a tensor has only one element, even if it's 1D?).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/mkldnn_lowerings.py [hunk 1]
```diff
@@ -889,6 +896,25 @@ def qlinear_binary(
...
+
+            if x_zp is None:
+                x_zp = V.graph.add_tensor_constant(
+                    torch.tensor(0, dtype=torch.int32), name="x_zp"
+                )
+
+            if w_zp is None:
+                w_zp = V.graph.add_tensor_constant(
+                    torch.tensor(0, dtype=torch.int32), name="w_zp"
+                )
```
Comment (sanchitintel) [inline torch/_inductor/mkldnn_lowerings.py:917]: This particular change is now necessary because of [an unnecessary change](https://github.com/pytorch/pytorch/pull/143187#discussion_r1921052467) introduced in https://github.com/pytorch/pytorch/pull/143187/commits/09bdb53dc849735714d767fd81d59b3c402da38f

File: torch/_inductor/mkldnn_lowerings.py [hunk 2]
```diff
@@ -588,6 +588,14 @@ def qlinear_unary(
                     torch.tensor(x_scale, dtype=torch.float32), name="x_scale"
                 )
             else:
+                x_scale_size = x_scale.get_size()
+                if len(x_scale_size) == 2:
+                    if x_scale_size[0] == 1 and x_scale_size[1] == 1:
```
Comment (leslie-fang-intel) [inline torch/_inductor/mkldnn_lowerings.py:593]: Can we generalize this check as all the value inside `x_scale_size` is 1? and how about the case when `x_scale_size` is `[1, 1, 1]`?

File: torch/_inductor/mkldnn_lowerings.py [hunk 3]
```diff
@@ -894,7 +902,26 @@ def qlinear_binary(
                     torch.tensor(x_scale, dtype=torch.float32), name="x_scale"
                 )
             else:
+                x_scale_size = x_scale.get_size()
+                if len(x_scale_size) == 2:
```
Comment (leslie-fang-intel) [inline torch/_inductor/mkldnn_lowerings.py:906]: Do we have the UT to cover this case for `qlinear_binary`?

File: torch/_inductor/mkldnn_lowerings.py (hunk 2 above)
Comment (jansel) [inline torch/_inductor/mkldnn_lowerings.py:593]:
```suggestion
            if all(i==1 for in x_scale.get_size()):
```

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/mkldnn_lowerings.py (hunk 3 above)
Comment (sanchitintel) [reply torch/_inductor/mkldnn_lowerings.py:906]: @leslie-fang-intel @chunyuan-w, to clarify, the change here is for fixing `qlinear_binary` lowering, which is unrelated to GEMM auto-tuning. So, it's not appropriate for me to add a UT for it in `test_cpu_select_algorithm.py`, since the GEMM template is currently not working for `qlinear_binary` with sum post-op right now. I'll add a UT in `test_mkldnn_pattern_matcher.py` instead. Thanks!

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 4]
```diff
@@ -3644,6 +3646,8 @@ def forward(self, a):
                 c = c * self.b_scale
                 if self.has_bias:
                     c = c + self.bias
+                if inplace_add:
+                    c.add_(self.additive)
                 return c
 
         mod = Mod(dtype, has_bias).eval()
```
Comment (leslie-fang-intel) [inline test/inductor/test_mkldnn_pattern_matcher.py:3653]: If `inplace_add` is True, do we need to check that `qlinear_binary` fusion has been hit?

**Review by leslie-fang-intel — CHANGES_REQUESTED** [review]

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 4 above)
Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3688]: For these UTs, the Inductor pattern-matching passes in quantization.py are not responsible for lowering to `qlinear_binary`, so I couldn't use a corresponding counter, and added a check for the generated code instead.

BTW, here are some findings unrelated to the specific fix in this PR:
 
 1. `qlinear_binary` is not being used with the UTs when `M` dimension is not 1. I don't know if that's intentional. cc @Xia-Weiwen 
 2. An Inductor bug got exposed for dynamic shapes - when dynamic shapes are enabled and `x2` is added with a sum post-op in `qlinear_binary`, the outermost dim's stride of the output tensor can't be determined (The comments in code have a typo & state `innermost` instead of `outermost`). Details below.
 
 <details>

```python
# s1 below is undetermined
def call(args):
    arg4_1, arg5_1 = args
    args.clear()
    s0 = arg4_1
    assert_size_stride(arg5_1, (1, s0), (s0, 1))
    # Topologically Sorted Source Nodes: [add], Original ATen: [aten.add]
    torch.ops.onednn.qlinear_pointwise.binary_tensor(arg5_1, reinterpret_tensor(_frozen_param1, (), (), 0), x_zp, _frozen_param4, _frozen_param5, w_zp, _frozen_param3, None, 1.0, 0, torch.float32, 1.0, 0, 'sum', 1.0, 'none', [None], '')
    del arg5_1
    return (reinterpret_tensor(_frozen_param3, (1, 64), (s1, 1), 0), )
```

</details>

Comment (Xia-Weiwen) [reply test/inductor/test_mkldnn_pattern_matcher.py:3688]: > qlinear_binary is not being used with the UTs when M dimension is not 1.

You may add cases for that I think. I don't think we have restrictions on M for qllinear_binary.

Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:3688]: Sorry, I meant when M > 1, then lowering to `qlinear_binary` is not done

**Comment by sanchitintel** [comment]
Hi @leslie-fang-intel, thanks for adding the `trunk` label!
Looks like I'd have to disable the UT for ARM CPU based machines

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 5]
```diff
@@ -3657,7 +3696,10 @@ def matcher_check_fn():
         self._test_common(
             mod,
             (a,),
-            matcher_check_fn=matcher_check_fn,
+            matcher_check_fn,
+            include_ops=("qlinear_pointwise.binary_tensor",)
```
Comment (leslie-fang-intel) [inline test/inductor/test_mkldnn_pattern_matcher.py:3700]: Why don't we just check the `counters["inductor"]["qlinear_binary_matcher_count"]`?

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 6]
```diff
@@ -3632,18 +3661,28 @@ def __init__(self, dtype: torch.dtype, has_bias: bool):
...
+                if expand_a_scale:
+                    a_scale = self.a_scale.expand(c.shape)
+                else:
+                    a_scale = self.a_scale
                 c = c * a_scale
                 c = c * self.b_scale
                 if self.has_bias:
                     c = c + self.bias
+                elif inplace_add and test_for_pointwise_binary:
+                    # When M is 1, dynamic shapes are enabled with torch.compile, has_bias is False,
```
Comment (leslie-fang-intel) [inline test/inductor/test_mkldnn_pattern_matcher.py:3682]: When M is 1, I think Inductor will always do the specialization instead of using the symbolic size.

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 5 above)
Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py]: Now that I think of it. I should've used the counter.

For `[M=32, K=32, N=64]`, this pattern wasn't hit (and not even when I changed `K` to 48, it wasn't, so the counter was 0), and I thought this lowering wasn't because of those FX patterns. When I later discovered that only the `M=1` case was being lowered to `qlinear_binary`, I should've re-added the counter check.

File: test/inductor/test_cpu_select_algorithm.py [hunk 7]
```diff
@@ -1444,7 +1445,6 @@ def test_da8w8_sym_act_sym_wgt_with_int_mm(
         """
         if dtype == torch.bfloat16 and not torch.ops.mkldnn._is_mkldnn_bf16_supported():
             return
-        M = 32
         in_feature = 48
         out_feature = 64
         q_min, q_max = -32, 31
```
Comment (leslie-fang-intel) [inline test/inductor/test_cpu_select_algorithm.py:1450]: It seems this unit test doesn't verify whether the CPP Template has been used. Can we add a counter check similar to https://github.com/pytorch/pytorch/pull/147895?

Comment (sanchitintel) [reply test/inductor/test_cpu_select_algorithm.py:1450]: I could rebase this PR when #147895 will be merged (but it seems like its merge procedure ran into some CI issue), so I'll add it now.

File: test/inductor/test_cpu_select_algorithm.py [hunk 8]
```diff
@@ -1498,6 +1498,9 @@ def forward(self, a):
 
         vec_amx = VecAMX()
         self._check_amx_counter(vec_amx)
+        if torch._C._cpu._is_amx_tile_supported():
+            # Only AMX ISA based micro-kernel is currently supported for da8w8
+            self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
```
Comment (leslie-fang-intel) [inline test/inductor/test_cpu_select_algorithm.py:1503]: I think we should use the new counters add by you recently, right?

Comment (sanchitintel) [reply test/inductor/test_cpu_select_algorithm.py:1503]: No, because #147895 didn't get merged (timed out due to non-completion of some CI jobs)

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 6 above)
Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:4089]: Tracking that issue at #148199. Thanks!

File: torch/_inductor/mkldnn_lowerings.py (hunk 2 above)
Comment (sanchitintel) [reply torch/_inductor/mkldnn_lowerings.py]: Thank you! I applied this change at a couple of places.

**Review by jansel — APPROVED** [review]

**Review by leslie-fang-intel — APPROVED** [review]
